GDB MI Adapter for OpenOCD Embedded Debugging in Binary Ninja

anvol opened the pull request:
## __Key Features:__

- __OpenOCD Integration__: Seamless connection to OpenOCD debug servers (default port 3333)
- __ARM Cortex-M Focus__: Optimized for Cortex-M3, M4, M7, and other embedded ARM variants
- __Architecture Auto-Detection__: Intelligent detection of target architecture through register analysis and GDB queries
- __Thread-Safe Design__: Non-blocking UI with dedicated reader threads and command synchronization
- __Debugging__: Breakpoints, stepping, memory access, register inspection, and watch variables/structs in bn views

### Technical Implementation Details

__Architecture:__

- __GdbMiConnector__: Handles low-level MI protocol communication, process management, and async event processing. Covered with `miparser_test`
- __GdbMiAdapter__: Implements Binary Ninja's DebugAdapter interface with high-level debugging operations
- __Caching System__: Performance-optimized caching for threads, registers, and stack frames
- __Cross-Platform__: Proper process handling Linux, macOS, and (maybe) Windows

__Key Technical Components:__

- MI protocol parser handling nested structures and complex responses
- Variable object system for efficient watch variable tracking
- Hardware breakpoint support for embedded targets
- Memory access constraints for embedded memory maps

### Setup and Configuration

__Required Tools:__

- __Host System__: Linux or macOS with Binary Ninja
- __Debug Probe__: ST-Link, J-Link, or other OpenOCD-compatible probe
- __GDB__: `gdb-multiarch` or `arm-none-eabi-gdb`
- __OpenOCD__: Running with target configuration (e.g., STM32F4x)

__Typical Workflow:__

1. Start OpenOCD: `openocd -f interface/stlink.cfg -f target/stm32f4x.cfg`
2. Export DWARF file for target architecture
3. Configure adapter settings in Binary Ninja. Specify desired GDB path, connection IP/port and DWARF file
4. Connect to target through the debugger interface
5. Set breakpoints, inspect memory, and step through firmware code

__Adapter Settings:__

- `gdb.path`: Path to GDB executable (default: `gdb-multiarch`)
- `connect.ipAddress`: OpenOCD server IP (default: `127.0.0.1`)
- `connect.port`: OpenOCD GDB port (default: `3333`)
- `common.inputFile`: Firmware file for base address resolution
- `gdb.symbolFile`: ELF or DWARF debug information so gdb would know funcs boundaries and names for backtrace

### Testing and Validation

__Verified Platforms:__

- Debian 13 with gdb-multiarch
- macOS 15+ (Apple Silicon) arm-none-eabi-gdb from brew
- ARM Cortex-M targets (STM32F103 and STM32F401 extensively tested)

__Validation Checklist:__

- [x] GDB process launches and terminates correctly
- [x] Architecture auto-detection for ARM Cortex-M targets
- [x] Successful connection to OpenOCD
- [x] Register display and memory access
- [x] Breakpoint setting and hitting
- [x] Stepping and control flow
- [x] Watch variable updates in main ui window
- [x] Thread and stack frame information

<img width="1591" height="950" alt="image" src="https://github.com/user-attachments/assets/247cdbb1-1488-4b15-b3a4-e7df7d9084f8" />

<img width="1840" height="1191" alt="image" src="https://github.com/user-attachments/assets/a3a4c415-04e8-4a62-93b6-932f8aec54de" />

<img width="897" height="448" alt="image" src="https://github.com/user-attachments/assets/c681c460-88a4-4d5b-9a29-b220a1180d69" />


### Benefits for Binary Ninja Community

This implementation significantly enhances Binary Ninja's capabilities for:

1. __Embedded Security Research__: Dynamic analysis of IoT devices, automotive systems, and industrial controllers
2. __Firmware Reverse Engineering__: Live debugging of embedded binaries with Binary Ninja's advanced analysis features
3. __Hardware-Assisted Analysis__: Integration with real hardware through standard debugging probes
4. __Educational Use__: Teaching embedded systems and reverse engineering with professional tools

anvol committed:
initial implementation of GDB MI with basic setup for stm32f4
GDB MI now implements most basic functionality
Notify UI that we're connected and ready to serve data
small fixes to handle quit
cleanup
feat(gdb): Refactor event handling and command execution
- An event loop (`EventLoop`) running in a dedicated thread (`m_eventThread`) now handles all asynchronous output from GDB. This replaces the previous callback-based approach (`AsyncRecordHandler`).
- A mutex (`m_gdbCommandMutex`) is now used exclusively within the `SendCommand` method, centralizing locking and preventing race conditions. Higher-level functions no longer manage the lock themselves.
- The target's running state is now managed by a `std::atomic<bool>` (`m_targetRunningAtomic`) for safe access across threads.
- Asynchronous commands (`-exec-run`, `-exec-step`, etc.) are now sent via a new `SendCommandAsync` method, which does not wait for a result, allowing the UI to remain responsive while the target is running.
- Logging has been adjusted, changing many `LogInfo` calls to `LogDebug` to reduce default output verbosity.
- The GDB-MI value parser (`MiValue::parse`) has been rewritten to be more robust and correctly handle complex GDB output formats.
Add watch widget implementation for basic types
MIParser fixes and cleanup
Added unit test for MI parser
Implemented StepOut function

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Vector35/debugger?pullRequest=875) <br/>All committers have signed the CLA.

anvol commented:
Implements #874 and partially covers #170

anvol committed:
- clear register names before populating
- fix missing default init values

anvol committed:
fix bp log message

anvol committed:
Improve ReaderThread exit handling
Clear cache on stop/quit/restart

xusheng6 commented:
Thanks for the contribution! I had a quick glance and it looks good, I will review it when I get time. 

Also I had copilot look at it as well -- I have not gone through them yet, if they are correct, please address them; if they are wrong, please point it out!

anvol commented:
- buffer - confirmed, fixed that locally
- step - confirmed
- additionally, GetBreakpointList should be implemented as GDB MI expects it's idx from "-break-list" cmd to remove breakpoint "-break-delete <id>"

anvol committed:
Fix StepOver
Remove all breakpoints at address
Use BackendMessageEventType as output for gdb messages

xusheng6 commented:
I forked your PR and created https://github.com/Vector35/debugger/tree/add_gdb_mi_adapter

I noticed that some of the things you put are embedded specific, which I hope to remove before merging. I also noted that they might still be useful for you, so I am starting on top of a new branch

xusheng6 commented:
@anvol I tried to make this branch into a state that can be merged into the dev branch, but I failed to do so. I met some issues when using the GDB MI to connect to a gdbserver running on Linux and getting the list of loaded modules. I know this is beyond your interest, but I believe this needs to be addressed before I can ship it because many users would try to do that anyways.

This PR/branch has missed our internal feature freeze, which means it would not be in stable 5.2. I will try to get back to this after the release!

xusheng6 commented:
The PR is manually merged into https://github.com/Vector35/debugger/commit/052139fc0bd04e00a6878487a16e298bdf32f765, thanks a lot for the contribution! @anvol
